src/compiler/parser.ts

```diff
@@ -3569,18 +3569,51 @@ namespace ts {
             return parsePostfixTypeOrHigher();
         }
 
+        function parseFunctionOrConstructorTypeToError(
+            isInUnionType: boolean
+        ): TypeNode | undefined {
+            // the function type and constructor type shorthand notation
+            // are not allowed directly in unions and intersections, but we'll
+            // try to parse them gracefully and issue a helpful message.
+            if (isStartOfFunctionType() || token() === SyntaxKind.NewKeyword) {
+                const type = parseFunctionOrConstructorType();
+                let diagnostic: DiagnosticMessage;
+                if (isFunctionTypeNode(type)) {
+                    if (isInUnionType) {
+                        diagnostic = Diagnostics.Function_type_notation_must_be_parenthesized_when_used_in_a_union_type;
+                    }
+                    else {
+                        diagnostic = Diagnostics.Function_type_notation_must_be_parenthesized_when_used_in_an_intersection_type;
+                    }
+                }
+                else {
+                    if (isInUnionType) {
+                        diagnostic = Diagnostics.Constructor_type_notation_must_be_parenthesized_when_used_in_a_union_type;
+                    }
+                    else {
+                        diagnostic = Diagnostics.Constructor_type_notation_must_be_parenthesized_when_used_in_an_intersection_type;
+                    }
+                }
+                parseErrorAtRange(type, diagnostic);

```

Comment:
side note: I bet this newish function could replace a lot of other lower-level error-reporting functions in the parser; it's only used 3 other places right now.